homeassistant/components/proxmoxve/sensor.py

```diff
@@ -0,0 +1,127 @@
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from . import COORDINATORS, DOMAIN, PROXMOX_CLIENTS, ProxmoxEntity
+
+BYTE_TO_GIBIBYTE = 1.074e9
+
+
+@dataclass
+class ProxmoxVERequiredKeyMixin:
+    """Mixin for required lambda functions."""
+
+    native_lambda: Callable
+
+
+@dataclass
+class ProxmoxVESensorEntityDescription(
+    SensorEntityDescription, ProxmoxVERequiredKeyMixin
+):
+    """Class to describe a proxmoxve sensor."""
+
+
+SENSOR_DESCRIPTIONS = [
+    (
+        ProxmoxVESensorEntityDescription(
+            key="mem_gib",
+            name="mem_gib",
+            native_unit_of_measurement=DATA_GIBIBYTES,
+            icon=None,
+            native_lambda=lambda data: round(int(data.mem) / BYTE_TO_GIBIBYTE, 2),
+        )
+    ),
+]
+
+
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None,
+) -> None:
+    """Set up sensors."""
+
+    if discovery_info is None:
+        return
+
+    sensors = []
+
+    for host_config in discovery_info["config"][DOMAIN]:
+        host_name = host_config["host"]
+        host_name_coordinators = hass.data[DOMAIN][COORDINATORS][host_name]
+
+        if hass.data[PROXMOX_CLIENTS][host_name] is None:
+            continue
+
+        for node_config in host_config["nodes"]:
+            node_name = node_config["node"]
+
+            for dev_id in node_config["vms"] + node_config["containers"]:
+                coordinator = host_name_coordinators[node_name][dev_id]
+
+                # unfound case
+                if (coordinator_data := coordinator.data) is None:
+                    continue
+
+                name = coordinator_data.name
+
+                for description in SENSOR_DESCRIPTIONS:
+                    sensors.append(
+                        ProxmoxSensor(
+                            coordinator=coordinator,
+                            vm_name=name,
+                            host_name=host_name,
+                            node_name=node_name,
+                            vm_id=dev_id,
+                            description=description,
+                        )
+                    )
+
+    add_entities(sensors)
+
+
+class ProxmoxSensor(ProxmoxEntity, SensorEntity):
+    """A sensor for reading Proxmox VE data."""
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        vm_name: str,
+        host_name: str,
+        node_name: str,
+        vm_id: int,
+        description: ProxmoxVESensorEntityDescription,
+    ) -> None:
+        """Create the sensor for vms or containers."""
+
+        self.native_lambda = description.native_lambda
+        self._attr_native_unit_of_measurement = description.native_unit_of_measurement
+        super().__init__(
+            coordinator,
+            unique_id=f"proxmox_{node_name}_{vm_id}_{description.name}",
```

Comment:
Entities should not include the domain name: https://developers.home-assistant.io/docs/entity_registry_index?_highlight=unique_id#unique-id

Also, hostname is not an acceptable source for a unique id as described there in the docs
VM ID is not an acceptable source for a unique id either as they come and go as well.

I realize these are copied from the existing pattern, but i think those aren't quite right, and need to be migrated.

Reply:
I'm not sure what should be used then if VM ID can't be used. VM IDs are unique to proxmox hosts, although different hosts can have the same id and node name unfortunately.

 What would you suggest be used for unique ids for proxmox hosts, which there can be multiple of in one configuration?

Reply:
This page has all the guidance: https://developers.home-assistant.io/docs/entity_registry_index/#unique-id

Unique IDs awesome, but i did not get the impression the are required and you should use "unacceptable sources for a unique ID" anyway even if you don't have one...

Reply:
Could we use the proxmox node name as a prefix, followed by the VM id?

Looking through proxmox API documentation ([great tool for exploring here](https://pve.proxmox.com/pve-docs/api-viewer/index.html#/nodes/{node})), the only unique value I can think of is the self signed certificate's (default cert) hash for the node. The MAC isn't exposed on the `interfaces` endpoint, and there isn't anything else unique to use that can be assumed is provided by all Proxmox hosts.

Reply:
I used the following pattern in this PR #5247 which is still under review:
https://github.com/home-assistant/core/blob/b1207badb42ca39b88a019e5e5f77cfa8eb9ee12/homeassistant/components/proxmoxve/binary_sensor.py#L85

The goal is to implement more features (like this PR) after adding the configuration flow (to attend [ADR010](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md)).
I have the [custom component](https://github.com/dougiteixeira/proxmoxve) (beta version) is still in the testing period, with the configuration flow and the sensors/commands implemented.

Reply:
> I used the following pattern in this PR #75247 which is still under review:
> 
> https://github.com/home-assistant/core/blob/b1207badb42ca39b88a019e5e5f77cfa8eb9ee12/homeassistant/components/proxmoxve/binary_sensor.py#L85

**My take on the unique ID composition:**

When it comes to a virtual environment, it is difficult to cling to something immutable like a serial number, MAC Address or something physically recorded.

The parameters that were considered to generate a unique ID are hardly changed after a Proxmox instance goes into production.

Thinking about a planned structure and already in operation, even if the hardware is replaced, the structure's configurations in theory remain unchanged (host, port, ID of nodes and virtual machines, etc…) and are migrated with backup.

Parameters considered:

* Host and port: defines a single destination on the network, it is not allowed to change it in the configuration entry, if there is a change, the configuration entry must be deleted and a new one created with the correct data. Using the port is a way to not restrict who uses network port forwarding (NAT) to more than one server on the same network.
* Node: for a Proxmox installation this is the node identifier on the server, it can be changed in the Proxmox settings, but this would represent a new node for the integration. Changing the configuration entry will not be allowed, if the change occurs on the server, the configuration entry must be deleted and a new one with the correct data created.
* VM ID: is the identifier that Proxmox (API) uses for all tasks involving the virtual machine/container. It is even possible to change it in Proxmox, but this involves more advanced file manipulation. So a different ID would be a new VM/CT for the integration, as the resource would no longer be available with that identifier, so the device would have to be removed and the new one added (via the config entry options flow).

Reply:
This makes sense to me. So this PR is effectively blocked then until the refactor from #75247 is merged? Then @Corbeno could rebase and use that unique_id scheme

Reply:
> This makes sense to me. So this PR is effectively blocked then until the refactor from #75247 is merged? Then @Corbeno could rebase and use that unique_id scheme

I'm not sure if it will be blocked, because ADR010 doesn't allow changes to the YAML configuration, about new features like the Sensor platform I'm not sure...

Reply:
Considering that the configuration flow is being implemented and that it is unique for host/port/node, we can simplify and use the unique id as `f"{config_entry.entry_id}_{vm_id}_{key}"`.

In this way, we will comply with the documentation that restricts the use of hostname and allows the use of the Config Entry ID as a last resort.

Reply:
If you want to unblock the fastest way is to not set a unique Id right? That was what I was trying to say above about unique ids not being a requirement. You can unblock and try to rally for support around a particular solution in a follow up PR.

Reply:
The host might be identified by its primary nic mac address, for the VM it's still the best solution to use its ID.

As long as the environment is virtual you cannot refer to a piece of hardware.

You can maybe use a description to set a custom identifier but it can also be wiped 

This is more an issue with proxmox than HA since you can have conflicts when moving a VM from one cluster to another

Reply:
I got a second opinion and it seems like we're ok with using a vmid, since the only way it changes is if the user changes it.

Reply:
Good point, I was mistaken thinking we were in a world where this was using config entries. 

VM ids are unique within a cluster. Node names are not really needed at all -- in a multi-node cluster, the node name is not really used anywhere. you'd communicate with the host and do work in terms of the vm id.

It seems like really the unique id needs to be host + vmid?   Now I realize that you could technically configure two hosts within the same cluster here and get duplicate vmids, but that seems part to prevent without moving this to config flows.

Am I understanding your point?